clientd3d/d3drender_bgoverlays.c

```diff
@@ -0,0 +1,310 @@
+		int height_max = 200;
+		int height_min = -200;
+
+		long object_width = DibWidth(pDib) * size_scaler;
+		long object_height = DibHeight(pDib) * size_scaler;
+
+		d3d_render_packet_new* pPacket = D3DRenderPacketFindMatch(pPool, NULL, pDib, 0, 0, 0);
+		if (NULL == pPacket)
+			return;
+
+		d3d_render_chunk_new* pChunk = D3DRenderChunkNew(pPacket);
+		assert(pChunk);
+
+		pPacket->pMaterialFctn = &D3DMaterialWorldPacket;
+
+		pChunk->numIndices = 4;
+		pChunk->numVertices = 4;
+		pChunk->numPrimitives = pChunk->numVertices - 2;
+		pChunk->pMaterialFctn = &D3DMaterialWorldDynamicChunk;
+		pChunk->flags |= D3DRENDER_NOAMBIENT | D3DRENDER_WORLD_OBJ;
+		pChunk->zBias = ZBIAS_BASE;
+
+		int piAngle = overlay->x;
+		int piHeight = overlay->y;
+
+		auto mapRange = [](double value, double in_min, double in_max, double out_min, double out_max) -> double {
+			return (value - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
+		};
+
+		long mappedHeightValue = mapRange(piHeight, -200, 200, height_min, height_max);
+		float angleInRadians = piAngle / 4096.0f;
+		double azimuthalAngle = angleInRadians * 2 * PI;
+		long radius = height_max * 5;
+		double horizontalRadius = sqrt(pow(radius, 2) - pow(mappedHeightValue, 2));
+
+		long x = params->viewer_x - (object_width / 2) + horizontalRadius * cos(azimuthalAngle);
+		long y = params->viewer_y - (object_height / 2) + horizontalRadius * sin(azimuthalAngle);
+		long z = params->viewer_height + mappedHeightValue;
+
+		Vector3D bg_overlay_pos;
+		bg_overlay_pos.x = x;
+		bg_overlay_pos.y = y;
+		bg_overlay_pos.z = z;
+
+		D3DMATRIX rot, mat;
+		MatrixIdentity(&mat);
+		MatrixIdentity(&rot);
+
+		const float FULL_CIRCLE_TO_DEGREES = 360.0f / 4096.0f;
+		const float DEGREES_TO_RADIANS = PI / 180.0f;
+		const float ANGLE_RANGE_TO_DEGREES = 45.0f / 414.0f;
+
+		MatrixRotateY(&rot, (float)angleHeading * FULL_CIRCLE_TO_DEGREES * DEGREES_TO_RADIANS);
+		MatrixRotateX(&mat, (float)anglePitch * ANGLE_RANGE_TO_DEGREES * DEGREES_TO_RADIANS);
+		MatrixTranspose(&rot, &rot);
+		MatrixTranslate(&mat, bg_overlay_pos.x, bg_overlay_pos.z, bg_overlay_pos.y);
+		MatrixMultiply(&pChunk->xForm, &rot, &mat);
+
+		pChunk->xyz[0].x = 0;
+		pChunk->xyz[0].z = 0;
+		pChunk->xyz[0].y = 0;
+
+		pChunk->xyz[1].x = -object_width;
+		pChunk->xyz[1].z = 0;
+		pChunk->xyz[1].y = 0;
+
+		pChunk->xyz[2].x = -object_width;
+		pChunk->xyz[2].z = object_height;
+		pChunk->xyz[2].y = 0;
+
+		pChunk->xyz[3].x = 0;
+		pChunk->xyz[3].z = object_height;
+		pChunk->xyz[3].y = 0;
+
+		for (int j = 0; j < 4; j++) {
+			pChunk->bgra[j].b = 255;
+			pChunk->bgra[j].g = 255;
+			pChunk->bgra[j].r = 255;
+			pChunk->bgra[j].a = 255;
+		}
+
+		static bool initialized = false;
+		static float u[4] = { 0.0f, 0.0f, 1.0f, 1.0f };
+		static float v[4] = { 0.0f, 1.0f, 1.0f, 0.0f };
+		const float epsilon = 0.007f; // Small value to adjust UV coordinates inward
+		if (!initialized)
+		{
+			// Set the texture coordinate template with epsilon adjustment once.
```

Comment:
This is super weird (I realize it already existed).. should just initialize u and v with the correct values, make them const, and get rid of the "initialized" variable.